theano/tensor/opt_uncanonicalize.py

```diff
@@ -98,3 +99,45 @@ def local_max_to_min(node):
                                  max.owner.op.axis)(neg.owner.inputs[0])]
 
     return False
+
+
+@register_uncanonicalize
+@gof.local_optimizer([T.Alloc])
+def local_alloc_dimshuffle(node):
+    """
+    If a dimshuffle is inside an alloc and only adds dimension to the
+    left, remove it.
+    """
+    if node.op == T.alloc:
+        input_ = node.inputs[0]
+        if input_.owner and isinstance(input_.owner.op, DimShuffle):
+            # check if it only adds dimension to the left
+            new_order = input_.owner.op.new_order
+            expected_new_order = ('x',) * (input_.ndim - input_.owner.inputs[0].ndim) + \
+                    tuple(range(input_.owner.inputs[0].ndim))
+            if new_order != expected_new_order:
+                return False
+            return [T.alloc(input_.owner.inputs[0], *node.inputs[1:])]
+    return False
+
+
+@register_uncanonicalize
+@gof.local_optimizer([T.Reshape])
+def local_reshape_dimshuffle(node):
+    """
+    If a dimshuffle is inside a reshape and does not change the order
+    of dimensions, remove it.
+    """
+    if isinstance(node.op, T.Reshape):
+        input_ = node.inputs[0]
+        if input_.owner and isinstance(input_.owner.op, DimShuffle):
+            new_order = input_.owner.op.new_order
+            offset = 0
+            for i, dim in enumerate(new_order):
+                if dim == 'x':
+                    offset += 1
+                    continue
+                elif i != dim + offset:
+                    return False
```

Comment:
The entire if don't do what we want. I think it should be:
if dim == 'x':
    continue
elif i != dim + offset:
   return
else:
    offset += 1

THis will make sure we don't swap dimensions and allow 'x' anywhere.